File: py/2022/day09.py
import itertools
from math import copysign

from aoc import main
from aoc.coords import Dir, Point, Vector, addp, subp


def parse(s: str) -> list[Vector]:
    rv = []
    for line in s.splitlines():
        dir, count = line.split(" ")
        rv.extend([Dir.parse(dir)] * int(count))
    return rv


def sign(v: Vector) -> Vector:
    return int(copysign(1 if v[0] else 0, v[0])), int(copysign(1 if v[1] else 0, v[1]))


def follow(t: Point, h: Point) -> Point:
    delta = subp(h, t)
    return t if all(abs(x) < 2 for x in delta) else addp(t, sign(delta))


def walk(moves: list[Vector], n: int):
    origin = (0, 0)
    visits = itertools.accumulate(moves, addp, initial=origin)
    for _ in range(n - 1):
        visits = itertools.accumulate(visits, follow, initial=origin)
    return len(set(tuple(p) for p in visits))


if __name__ == "__main__":
    main(
        lambda s: walk(parse(s), 2),
        lambda s: walk(parse(s), 10),
    )

File: ts/2022/day09.ts
import { main } from 'lib/advent';
import { Dir, move, Point, pointHash } from 'lib/coords';
import { iter, range } from 'lib/iter';
import { lines } from 'lib/util';

const dirMap = {
  U: Dir.Up,
  D: Dir.Down,
  L: Dir.Left,
  R: Dir.Right,
};

function parse(lines: string[]): Dir[] {
  return lines.flatMap((line) => {
    const [dir, count] = line.split(' ');
    return Array(Number(count)).fill(dirMap[dir]);
  });
}

function follow(tail: Point, head: Point): Point {
  const dx = head.x - tail.x;
  const dy = head.y - tail.y;
  if (Math.abs(dx) < 2 && Math.abs(dy) < 2) return tail;
  return { x: tail.x + Math.sign(dx), y: tail.y + Math.sign(dy) };
}

function tailVisits(moves: Dir[], n: number): number {
  const start: Point = { x: 0, y: 0 };
  const headVisits = iter(moves).scan(move, start);
  return range(1, n)
    .reduce((visits) => visits.scan(follow, start), headVisits)
    .uniqBy(pointHash).size;
}

main(
  (s) => tailVisits(parse(lines(s)), 2),
  (s) => tailVisits(parse(lines(s)), 10)
);
